File: frontend/src/features/disasters/services/incidentService.ts
import * as incidentsApi from "../../../shared/api/incidents";
import * as archetypesApi from "../../../shared/api/archetypes";
import type { Disaster, IncidentArchetypeDsl } from "../types/disasters.types";
import type {
  ApiIncidentResponse,
  ApiIncidentCreate,
  ApiIncidentUpdate,
  ApiIncidentArchetypeResponse,
} from "../../../shared/api/types";

function mapIncident(api: ApiIncidentResponse): Disaster {
  const values = api.archetype_values as Record<string, unknown> | undefined;
  const restValues = values ? { ...values } : {};
  delete restValues.type;
  return {
    id: api.id,
    archetypeId: api.archetype_id || undefined,
    type: api.type,
    location: { lat: api.location_lat, lng: api.location_lng, address: api.location_address },
    address: api.location_address,
    severity: api.severity as Disaster["severity"],
    status: api.status as Disaster["status"],
    timestamp: api.timestamp,
    description: api.description,
    affectedRadius: api.affected_radius ?? undefined,
    archetypeValues: Object.keys(restValues).length > 0 ? restValues : undefined,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

function toApiCreate(item: Disaster): ApiIncidentCreate {
  return {
    archetype_id: item.archetypeId,
    type: item.type,
    location_lat: item.location.lat,
    location_lng: item.location.lng,
    location_address: item.location.address,
    severity: item.severity,
    status: item.status,
    timestamp: item.timestamp,
    description: item.description,
    affected_radius: item.affectedRadius,
    archetype_values: {
      ...(item.archetypeValues || {}),
    },
  };
}

function toApiUpdate(item: Disaster): ApiIncidentUpdate {
  return {
    archetype_id: item.archetypeId,
    type: item.type,
    location_lat: item.location.lat,
    location_lng: item.location.lng,
    location_address: item.location.address,
    severity: item.severity,
    status: item.status,
    description: item.description,
    affected_radius: item.affectedRadius,
    archetype_values: {
      ...(item.archetypeValues || {}),
    },
  };
}

export async function fetchIncidents(params?: {
  status?: string;
  severity?: string;
  search?: string;
}): Promise<Disaster[]> {
  const entries = await incidentsApi.listIncidents(params);
  return entries.map(mapIncident);
}

export async function saveIncident(item: Disaster): Promise<Disaster> {
  const existing = await incidentsApi.listIncidents({ search: item.id });
  if (existing.length > 0) {
    const api = await incidentsApi.updateIncident(item.id, toApiUpdate(item));
    return mapIncident(api);
  }
  const api = await incidentsApi.createIncident(toApiCreate(item));
  return mapIncident(api);
}

export async function removeIncident(id: string): Promise<void> {
  await incidentsApi.deleteIncident(id);
}

function mapApiArchetypeToDsl(api: ApiIncidentArchetypeResponse): IncidentArchetypeDsl {
  return {
    id: api.id,
    name: api.name,
    description: api.description,
    category: api.category,
    source: api.source,
    fieldSchema: api.field_schema || [],
    urgencyRules: api.urgency_rules || [],
    implications: api.implications || {},
    defaultReportUrgency: api.default_report_urgency || undefined,
    parentArchetypeId: api.parent_archetype_id || undefined,
    version: api.version,
  };
}

export async function listIncidentArchetypes(): Promise<IncidentArchetypeDsl[]> {
  const allArchetypes = await archetypesApi.listArchetypes();
  return allArchetypes
    .filter((a) => a.type === "incident")
    .map((a) => ({
      id: a.id,
      name: a.name,
      description: "",
      category: a.category,
      source: a.source,
      fieldSchema: (a.field_schema || []).map((f) => ({
        field: f.field,
        label: f.label,
        type: f.type,
        required: false,
      })),
      urgencyRules: (a.urgency_rules || []).map((r) => ({
        field: r.field,
        operator: r.operator,
        value: r.value,
        setUrgency: r.setUrgency,
        message: r.message,
      })),
      implications: {},
      version: a.version,
    }));
}

export async function fetchIncidentArchetypeWithInheritance(
  archetypeId: string
): Promise<{ dsl: IncidentArchetypeDsl; chain: IncidentArchetypeDsl[] }> {
  const chain: IncidentArchetypeDsl[] = [];
  let currentId: string | undefined = archetypeId;

  while (currentId) {
    const api = await archetypesApi.getIncidentArchetype(currentId);
    const dsl = mapApiArchetypeToDsl(api);
    chain.push(dsl);
    currentId = dsl.parentArchetypeId;
  }

  const child = chain[0];
  const parent = chain.length > 1 ? chain[1] : null;

  const merged: IncidentArchetypeDsl = {
    ...child,
    fieldSchema: mergeFieldSchemas(parent?.fieldSchema || [], child.fieldSchema),
    urgencyRules: parent?.urgencyRules
      ? mergeUrgencyRules(parent.urgencyRules, child.urgencyRules)
      : child.urgencyRules,
    implications: {
      ...(parent?.implications || {}),
      ...child.implications,
    },
  };

  return { dsl: merged, chain };
}

function mergeFieldSchemas(
  parentFields: IncidentArchetypeDsl["fieldSchema"],
  childFields: IncidentArchetypeDsl["fieldSchema"]
): IncidentArchetypeDsl["fieldSchema"] {
  const fieldMap = new Map<string, IncidentArchetypeDsl["fieldSchema"][number]>();
  for (const f of parentFields) {
    fieldMap.set(f.field, f);
  }
  for (const f of childFields) {
    fieldMap.set(f.field, { ...fieldMap.get(f.field), ...f });
  }
  return Array.from(fieldMap.values());
}

function mergeUrgencyRules(
  parentRules: IncidentArchetypeDsl["urgencyRules"],
  childRules: IncidentArchetypeDsl["urgencyRules"]
): IncidentArchetypeDsl["urgencyRules"] {
  const ruleKey = (r: IncidentArchetypeDsl["urgencyRules"][number]) =>
    `${r.field}|${r.operator}|${r.value}`;
  const ruleMap = new Map<string, IncidentArchetypeDsl["urgencyRules"][number]>();
  for (const r of parentRules) {
    ruleMap.set(ruleKey(r), r);
  }
  for (const r of childRules) {
    ruleMap.set(ruleKey(r), r);
  }
  return Array.from(ruleMap.values());
}

File: backend/app/services/incident_service.py
from datetime import datetime, timezone
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import or_

from app.db.models import Incident


class IncidentService:

    @staticmethod
    def list_incidents(
        db: Session,
        status: Optional[str] = None,
        severity: Optional[str] = None,
        search: Optional[str] = None,
    ) -> list[Incident]:
        query = db.query(Incident)
        if status:
            query = query.filter(Incident.status == status)
        if severity:
            query = query.filter(Incident.severity == severity)
        if search:
            query = query.filter(
                or_(
                    Incident.type.ilike(f"%{search}%"),
                    Incident.location_address.ilike(f"%{search}%"),
                    Incident.description.ilike(f"%{search}%"),
                )
            )
        return query.order_by(Incident.timestamp.desc()).all()

    @staticmethod
    def get_incident(db: Session, incident_id: str) -> Optional[Incident]:
        return (
            db.query(Incident)
            .filter(Incident.id == incident_id)
            .first()
        )

    @staticmethod
    def create_incident(
        db: Session,
        data: dict,
    ) -> Incident:
        incident = Incident(
            archetype_id=data.get("archetype_id"),
            type=data["type"],
            location_lat=data["location_lat"],
            location_lng=data["location_lng"],
            location_address=data["location_address"],
            severity=data["severity"],
            status=data.get("status", "aktif"),
            timestamp=data["timestamp"],
            description=data["description"],
            affected_radius=data.get("affected_radius"),
            archetype_values=data.get("archetype_values", {}),
        )
        db.add(incident)
        db.commit()
        db.refresh(incident)
        return incident

    @staticmethod
    def update_incident(
        db: Session,
        incident_id: str,
        data: dict,
    ) -> Optional[Incident]:
        incident = IncidentService.get_incident(db, incident_id)
        if not incident:
            return None

        for field, value in data.items():
            if value is not None:
                setattr(incident, field, value)

        incident.updated_at = datetime.now(timezone.utc)

        db.commit()
        db.refresh(incident)
        return incident

    @staticmethod
    def delete_incident(db: Session, incident_id: str) -> bool:
        incident = IncidentService.get_incident(db, incident_id)
        if not incident:
            return False
        db.delete(incident)
        db.commit()
        return True
